### ISSUE TITLE
ECDSA Twin Multiplication reference version is incorrect

### ISSUE THREAD

**Issue by weaversa** [GaloisInc/saw-script#315]
The `ec_twin_mult` function gives the following incorrect result on the test vector provided in the NIST specification.

```
p384_ec_mul> ec_affinify(p384_field, ec_twin_mul (p384_point_ops, 0xa4ebcae5a665983493ab3e626085a24c104311a761b5a8fdac052ed1f111a5c44f76f45659d2d111a61b5fdd97583480, {x=0xfba203b81bbd23f2b3be971cc23997e1ae4d89e69cb6f92385dda82768ada415ebab4167459da98e62b1332d1e73cb0e, y=0x5ffedbaefdeba603e7923e06cdb5d0c65b22301429293376d5c6944e3fa6259f162b4788de6987fd59aed5e4b5285e45}, 0xafcf88119a3a76c87acbd6008e1349b29f4ba9aa0e12ce89bcfcae2180b38d81ab8cf15095301a182afbc6893e75385d, {x=0xaacc05202e7fda6fc73d82f0a66220527da8117ee8f8330ead7d20ee6f255f582d8bd38c5a7f2b40bcdb68ba13d81051, y=0x84009a263fefba7c2c57cffa5db3634d286131afc0fca8d25afa22a7b5dce0d9470da89233cee178592f49b6fecb5092}))
{x = 0x76719313ffe00a156510426cde4921bf67024b2fe767d0686adbf8f65fb4713617afc75bc1d587bbaa241059364dae8b,
 y = 0x2cadceae9345392c63397e30197bfa4e7599c7a4e6b10350a976a0734f979584d11f0e810e8f6ece1297c8e4f04f8648}
```

The `p384_ec_twin_mul` function gives the correct result.
```
p384_ec_mul> ec_affinify(p384_field, p384_ec_twin_mul (0xa4ebcae5a665983493ab3e626085a24c104311a761b5a8fdac052ed1f111a5c44f76f45659d2d111a61b5fdd97583480, {x=0xfba203b81bbd23f2b3be971cc23997e1ae4d89e69cb6f92385dda82768ada415ebab4167459da98e62b1332d1e73cb0e, y=0x5ffedbaefdeba603e7923e06cdb5d0c65b22301429293376d5c6944e3fa6259f162b4788de6987fd59aed5e4b5285e45}, 0xafcf88119a3a76c87acbd6008e1349b29f4ba9aa0e12ce89bcfcae2180b38d81ab8cf15095301a182afbc6893e75385d, {x=0xaacc05202e7fda6fc73d82f0a66220527da8117ee8f8330ead7d20ee6f255f582d8bd38c5a7f2b40bcdb68ba13d81051, y=0x84009a263fefba7c2c57cffa5db3634d286131afc0fca8d25afa22a7b5dce0d9470da89233cee178592f49b6fecb5092}))
{x = 0x917ea28bcd641741ae5d18c2f1bd917ba68d34f0f0577387dc81260462aea60e2417b8bdc5d954fc729d211db23a02dc,
 y = 0x1a29f7ce6d074654d77b40888c73e92546c8f16a5ff6bcbd307f758d4aee684beff26f6742f597e2585c86da908f7186}
```

**Comment by weaversa** [comment]
After making these three changes to `ec_twin_mult` I get the vector to pass.

```
//drop`{gv+4}(r)@0
drop`{gv+5}(r)@0
```

```
//h0' = [c05, c04, c03, c02, c01]
h0' = [c01, c02, c03, c04, c05]
```

```
//h1' = [c15, c14, c13, c12, c11]
h1' = [c11, c12, c13, c14, c15]
```

**Comment by atomb** [comment]
Thanks, @weaversa ! If you'd like to create a pull request with those changes, we'd be happy to merge it. Otherwise, we can make those changes in a little bit.

**Comment by weaversa** [comment]
Will you please do it? I'm creating my own ECDSA using the new `Z` type and was using `ref_ec_mul.cry` as a reference. I'm afraid that during this process I've monkeyed with my own `ref_ec_mul.cry` a bit too much to make a pull request. It would be nice for someone to double check my changes as well.

**Comment by atomb** [comment]
Yep, no problem!

**Comment by joehendrix** [comment]
Thanks for catching this.  I pushed the suggested fix in 3d878bedb9e47f13ab3d2bdf8052343c3af8e109, and added a bit more functionality for testing purposes.

It would be nice to build some automated quick-check tests, but we need a way to generate random affine and projective elliptic-curve points.
